cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -36,38 +37,91 @@ namespace dataset {
 /// \brief A ScanTask backed by a parquet file and a subset of RowGroups.
 class ParquetScanTask : public ScanTask {
  public:
-  ParquetScanTask(std::vector<int> row_groups, std::vector<int> columns_projection,
+  ParquetScanTask(std::vector<int> row_groups,
                   std::shared_ptr<parquet::arrow::FileReader> reader,
                   std::shared_ptr<ScanOptions> options,
                   std::shared_ptr<ScanContext> context)
-      : row_groups_(std::move(row_groups)),
-        columns_projection_(std::move(columns_projection)),
-        reader_(reader),
-        options_(std::move(options)),
-        context_(std::move(context)) {}
+      : reader_(reader), options_(std::move(options)), context_(std::move(context)) {
+    size_t i = 0;
+    row_groups_ = MakeFunctionIterator([row_groups, i](int** out) mutable {
+      if (i < row_groups.size()) {
+        *out = &row_groups[i++];
+      } else {
+        *out = nullptr;
+      }
+      return Status::OK();
+    });
+  }
 
   RecordBatchIterator Scan() {
+    auto options = options_;
+    auto reader = reader_;
+    auto batches_it = MakeMaybeMapIterator(
+        [reader, options](int* row_group, RecordBatchIterator* out) {
+          return ScanRowGroup(*row_group, reader, options, out);
+        },
+        std::move(row_groups_));
+    return MakeFlattenIterator(std::move(batches_it));
+  }
+
+ private:
+  static Status ScanRowGroup(int row_group,
+                             const std::shared_ptr<parquet::arrow::FileReader>& reader,
+                             const std::shared_ptr<ScanOptions>& options,
+                             RecordBatchIterator* out) {
+    auto metadata = reader->parquet_reader()->metadata()->RowGroup(row_group);
+    std::vector<int> columns_projection;
+    if (!InferColumnProjection(*metadata, options, &columns_projection).ok()) {
+      // fall back to trivial projection
+      columns_projection = internal::Iota(metadata->num_columns());
```

Comment:
I'd move the fallback into InferColumnProjection in order to distinguish from a real error (if any in the future).

Reply:
If column projection cannot be inferred (for whatever reason), shouldn't we always fall back to trivial projection? As you said, this is just an optimization